### ISSUE TITLE
[CRD] No-op conversion should not be the default conversion

### ISSUE THREAD

**Issue by mbohlool** [kubernetes/kubernetes#65790]
In designing new conversions for CRDs we realized we would have a problem on the downgrade path if we default to no-op conversion all the time (as we do in 1.11). Assume we add some other conversions, e.g. webhook conversion. User defines two version and a conversion webhook to convert v1 to v2, assuming v1 is the storage version. The schema for v1 and v2 could be completely different and a controller/client working with v1 can fail unexpectedly if we give it v2 object (or vice versa). Now if for some reason the master is downgraded to 1.11, the default conversion will kick in and v1 and v2 object will have the same schema which fail some clients/controllers in a hard-to-debug manner.

One possible solution is to add a Conversion Strategy and set it to no-op by default in 1.11.1:

```
...
spec:
  conversion:
    strategy: no-op
...
```

and set this strategy to something else if we introduced a new conversion like webhook:

```
...
spec:
  conversion:
    strategy: external
...
```

now if we switch back to 1.11, the strategy value `external` is unrecognized and API server should stop serving CRD with proper error messages/channel (e.g. a condition).

**Comment by sttts** [comment]
Won't we have the same problem whenever we add a conversion feature like a nother kind of declarative conversion? Our normal unmarshaling mechanism will drop the new conversion kind and the conversions therefore are incomplete on a rollback.

**Comment by sttts** [comment]
The same is true when using a CRD with a /status subresource in 1.11. After switching back to version without beta support for subresources, the semantics of the endpoints fundamentally changes. Maybe one can argue that this  case might be easier to debug than a wrong schema.

**Comment by liggitt** [comment]
If there's a shared field (discriminated union) that indicates the type of conversion, the older server can at least recognize it is in an unknown state and disallow writes that would persist bad data, or refuse to serve the custom resources altogether

**Comment by sttts** [comment]
> If there's a shared field (discriminated union) that indicates the type of conversion, the older server can at least recognize it is in an unknown state and disallow writes that would persist bad data, or refuse to serve the custom resources altogether

That's what @mbohlool suggests, isn't it? I just add that it is not enough to only follow this pattern for the highlevel type of conversion (no-op, webhook, declarative), but we have to be very carefull for any other extension of the conversion mechanism, i.e. to have this defensive discriminated union handling at a much finer grained level.

**Comment by liggitt** [comment]
Yes. Fortunately, this isn't a situation you can get into without action taken by the cluster administrator. Documentation should be clear that any new conversion methods should only be used once there is no possibility of downgrading to a version where they do not exist and are supported

**Comment by mbohlool** [comment]
I am trying to at least solve the problem for conversions. With this strategy field we can achieve that. Maybe another solution would be having a featureFlagMap that sets a bit for each feature we support in CRD. each API server validate this map on serving to make sure all set bits are supported otherwise refuse to serve it and set the status of the CRD to unservable. When a CRD added by user, the API server look into its config and set the flag for each "used" feature. That way if a CRD did not use "/status" for example, the master can safely downgraded to a version that does not have that feature but if the CRD uses "/status" and we downgrade a that API server, the serving path sees a flag that does not understand and will stop serving it with proper error messages.

The problem here is we default to no-op converter. If we don't serve status, a client will get an error message but in this case we blindly return an unconverted object. That's why I think we should patch this ASAP.

**Comment by sttts** [comment]
A bitfield won't help if the users creates a CRD object in an old cluster. The user can then create one version of objects and read them back in another version and get unexpected results.

I have the feeling we are trying to solve one special case of a much more generic problem here, namely subversions of kinds and handling of unknown fields.

**Comment by liggitt** [comment]
>  I think we should patch this ASAP

I don't think it is urgent. You should not start using a new CRD conversion feature until your apiservers are upgraded and you will not roll back. Even if we added a conversion type discriminator field, that wouldn't protect from dropping unknown conversion-related fields when setting status via an old API server.

Documentation should be clear that any new conversion methods should only be used once there is no possibility of downgrading to a version where they do not exist and are supported

**Comment by mbohlool** [comment]
> that wouldn't protect from dropping unknown conversion-related fields when setting status via an old API server.

Can you explain this a little more?

**Comment by sttts** [comment]
> Can you explain this a little more?

An apiserver roundtrips CRDs (like every other resource) during update. @liggitt refers to status updates because these happen without user intervention. If you now have conversion related fields in your CRD spec, even a status update of that CRD will let those new fields not roundtrip in an old apiserver (after rollback). So in any case you will have dataloss (in this case after moving forward again). 

In other words: if you rollback with object in your etcd which have fields not known after rollback, you have data loss. The wrong conversions is the least of your problems in this case.

**Comment by mbohlool** [comment]
That is true but we would only have problem if we roll back to 1.11.0 or 1.11.1 (assuming the solution for this goes into 1.11.2). Those versions do not have conversion.Strategy but anything after that will. That's why I think it is good to have it as soon as possible. If we wait till 1.12 for this, then it is not safe to rollback to any 1.11 version. earlier versions like .0 and .1 are most probably not used in production so we would cover most of the cases if we get the solution for this in 1.11.2.

**Comment by liggitt** [comment]
> If we wait till 1.12 for this, then it is not safe to rollback to any 1.11 version.

Even with an explicit conversion type field, it wouldn't be safe to roll back a cluster with CRDs using webhook conversion to 1.11.x, since status updates would drop the webhook-related conversion fields in the CRD object.

**Comment by mbohlool** [comment]
>  since status updates would drop the webhook-related conversion fields in the CRD object.
Does it also drop this new discriminator field?

Anyway this change is not necessary for 1.11 anymore as the webhook conversion is alpha in 1.12 but we can merge it to master as it is part of the webhook design and I feel it is a good first step also preventing problems in downgrading 1.13 to 1.12 (with alpha flag off). I think this part should not be featuregate protected.

**Comment by sttts** [comment]
> but we can merge it to master as it is part of the webhook design

Let's get a fully working end-to-end prototype PR ready first. This we can review and then sign off the proposal. After that we can split it up in natural steps and get into master. This will be one of those steps. I don't like to merge in half of a design before it is actually finished.

**Comment by mbohlool** [comment]
I am going to close this issue then. One last note is the discriminator field should not be controller by the feature gate so we would have a safe downgrade from 1.13 to 1.12.
